build-tools/xaprepare/xaprepare/Steps/Step_InstallDotNetSDKs.cs

```diff
@@ -0,0 +1,113 @@
+					Utilities.DeleteDirectory (packToRemove);
+			}
+
+			// Delete any unnecessary SDKs if they exist.
+			var sdkPath = Path.Combine (dotnetPath, "sdk");
+			if (Directory.Exists (sdkPath)) {
+				var installedSdks = Directory.EnumerateDirectories (sdkPath).Select (d => new DirectoryInfo (d));
+				foreach (var sdkToRemove in installedSdks.Where (s => s.Name != dotnetVersion && s.Name != dotnetPreviewVersion))
+					Utilities.DeleteDirectory (sdkToRemove.FullName);
+			}
+
+			if (!await InstallDotNetAsync (context, dotnetPath, dotnetVersion)) {
+				Log.ErrorLine ($"Installation of dotnet SDK {dotnetVersion} failed.");
+				return false;
+			}
+
+			if (!await InstallDotNetAsync (context, dotnetPath, dotnetPreviewVersion)) {
+				Log.ErrorLine ($"Installation of dotnet SDK {dotnetPreviewVersion} failed.");
+				return false;
+			}
+
+			SetVariables (dotnetPath);
+			return true;
+		}
+
+		async Task<bool> InstallDotNetAsync (Context context, string dotnetPath, string version)
+		{
+			if (Directory.Exists (Path.Combine (dotnetPath, "sdk", version))) {
+				Log.Status ($"dotnet SDK version ");
+				Log.Status (version, ConsoleColor.Yellow);
+				Log.StatusLine (" already installed in: ", Path.Combine (dotnetPath, "sdk", version), tailColor: ConsoleColor.Cyan);
+				return true;
+			}
+
+			Uri dotnetScriptUrl = Configurables.Urls.DotNetInstallScript;
+			string dotnetScriptPath = Path.Combine (dotnetPath, Path.GetFileName (dotnetScriptUrl.LocalPath));
+
+			if (File.Exists (dotnetScriptPath))
+				Utilities.DeleteFile (dotnetScriptPath);
+			Log.StatusLine ("Downloading dotnet-install...");
+
+			(bool success, ulong size, HttpStatusCode status) = await Utilities.GetDownloadSizeWithStatus (dotnetScriptUrl);
+			if (!success) {
+				if (status == HttpStatusCode.NotFound)
+					Log.ErrorLine ("dotnet-install URL not found");
+				else
+					Log.ErrorLine ("Failed to obtain dotnet-install size. HTTP status code: {status} ({(int)status})");
+				return false;
+			}
+
+			DownloadStatus downloadStatus = Utilities.SetupDownloadStatus (context, size, context.InteractiveSession);
+			Log.StatusLine ($"  {context.Characters.Link} {dotnetScriptUrl}", ConsoleColor.White);
+			await Download (context, dotnetScriptUrl, dotnetScriptPath, "dotnet-install", Path.GetFileName (dotnetScriptUrl.LocalPath), downloadStatus);
+
+			if (!File.Exists (dotnetScriptPath)) {
+				Log.ErrorLine ($"Download of dotnet-install from {dotnetScriptUrl} failed");
+				return false;
+			}
+
+			Log.StatusLine ($"Installing dotnet SDK '{version}'...");
+
+
+			if (Context.IsWindows) {
+				return Utilities.RunCommand ("powershell.exe", new string [] {
+					"-NoProfile", "-ExecutionPolicy", "unrestricted", "-file", dotnetScriptPath,
+					"-Version", version, "-InstallDir", dotnetPath, "-Verbose"
+				});
+			} else {
+				return Utilities.RunCommand ("bash", new string [] {
```

Comment:
`bash` is no longer the default shell on `macOS` (it now uses `zsh`). Running `sh` might be safer.

Reply:
`sh` didn't work on Linux unfortunately, as this option `set -o pipefail` is used in the script, which the `sh` on our Linux VMs didn't seem to support. We could separate this out to use `sh` on macOS and `bash` on Linux, or we can keep it as is assume that this won't cause issues anytime soon.

Reply:
Let's keep it as-is then. If it breaks then we'll fix it